src/work/WorkScheduler.cpp

```diff
@@ -0,0 +1,67 @@
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "work/WorkScheduler.h"
+#include "lib/util/format.h"
+#include "util/Logging.h"
+
+namespace stellar
+{
+WorkScheduler::WorkScheduler(Application& app, std::function<void()> callback)
+    : Work(app, std::move(callback), "work-scheduler", BasicWork::RETRY_NEVER)
+{
+}
+
+WorkScheduler::~WorkScheduler()
+{
+    // TODO (mlo) With proper shutdown implementation, this call
+    // won't be necessary as WorkScheduler will transition into
+    // failure state and `reset` will take care of clearing children
+    clearChildren();
+}
+
+std::shared_ptr<WorkScheduler>
+WorkScheduler::create(Application& app)
+{
+    auto work = std::shared_ptr<WorkScheduler>(new WorkScheduler(app));
+    work->crankWork();
+    return work;
+};
+
+void
+WorkScheduler::onWakeUp()
+{
+    if (mScheduled)
+    {
+        return;
+    }
+
+    std::weak_ptr<WorkScheduler> weak(
+        std::static_pointer_cast<WorkScheduler>(shared_from_this()));
+    mScheduled = true;
+    mApp.getClock().getIOService().post([weak]() {
+        auto self = weak.lock();
+        if (!self)
+        {
+            return;
+        }
+        self->mScheduled = false;
+        self->crankWork();
+        if (self->getState() == WORK_RUNNING)
+        {
+            self->wakeUp();
```

Comment:
`wakeUp` calls `setState`, then `onWakeUp`, then `mNotifyCallback` if it wasn't already running (although as mentioned in `BasicWork::wakeUp` I think this should be changed slightly). For `WorkScheduler`, this is equivalent to `setState` then `onWakeUp` since `mNotifyCallback` is always empty. But here you are checking if it is already `WORK_RUNNING`, in which case `wakeUp` is equivalent to `onWakeUp`. May as well just call `onWakeUp` here directly.